python_modules/dagster/dagster/_core/definitions/data_version.py

```diff
@@ -508,3 +529,11 @@ def _is_volatile(self, *, key: AssetKey) -> bool:
         else:
             deps = self.asset_graph.get_parents(key)
             return len(deps) == 0 or any(self._is_volatile(key=dep_key) for dep_key in deps)
+
+    def _get_asset_record(self, *, key: AssetKey) -> Optional["AssetRecord"]:
+        # When targeting an asset, fetch records for all of its ancestors in
+        # one shot as a performance optimization. The records for all ancestors
+        # may be needed in determining the status.
+        ancestors = self.asset_graph.get_ancestors(key)
```

Comment:
This is O(n_ancestors^2), which can cause problems in larger asset graphs (we end up calling _get_asset_record for all the ancestors, so each of those ancestors will need to get their set of ancestors).

Two possible fixes:
- a public CachingInstanceQueryer.has_cached_asset_record() method, which we can call before this to avoid the get_ancestors call if we already have the record (and therefore can skip all this other stuff)
- turning get_ancestors into a cached method

I think the second option is still O(N^2) though as the prefetch_asset_records thing will need to iterate through the list of ancestors, so I think the first option is the way to go

Reply:
>a public CachingInstanceQueryer.has_cached_asset_record() method, which we can call before this to avoid the get_ancestors call if we already have the record (and therefore can skip all this other stuff)

I implemented this.